### PR TITLE
powi(i32) function for Decimal implemented

### PR DESCRIPTION
**Pull request by r001**
Decimal::powi(&self, exp:i32) -> Decimal function is now implemented with approximation. This solution will also work once we make Decimal::SCALE mutable.

### CODE REVIEW

File: scrypto/src/math/decimal.rs [hunk 1]
```diff
@@ -170,6 +173,65 @@ impl Decimal {
                 }
             }
         }
+
+    }
+
+    pub fn powi(&self, exp: i32) -> Self {
+        let mut x: BigInt = self.0.clone().into();
+        if x.is_negative() {
+            panic!("powi is not supported for negative numbers");
```
Comment (iamyulong) [inline scrypto/src/math/decimal.rs:182]: Why not? :)

File: scrypto/src/math/decimal.rs [hunk 2]
```diff
@@ -170,6 +173,65 @@ impl Decimal {
...
+        }
+        let s = Self::SCALE;
+        let bytes2 = Self::BITS / 4;
+        let b: BigInt = 10i128.pow(s).into();
```
Comment (iamyulong) [inline scrypto/src/math/decimal.rs:186]: `10i128.pow(s)` may overflow

File: scrypto/src/math/decimal.rs [hunk 3]
```diff
@@ -170,6 +173,65 @@ impl Decimal {
...
+                }
+                n /= 2i8;
+            }
+        }   
+
+        if exp.is_negative() {
+            Self(big_int_to_i128(b.clone() * b / z))
+        } else {
+            Self(big_int_to_i128(z))
+        }
```
Comment (iamyulong) [inline scrypto/src/math/decimal.rs:234]: Personally, I feel like there are too many intermediary steps and rounding, which makes it super hard to describe the behaviour of the `powi` function and hard for me to verify the correctness.

````
(a * 10 ^ -18) ^ exp = a ^ exp * 10 ^ (-18 * exp) 
                     = a ^ exp * 10 ^ (-18 * exp + 18) * 10 ^ -18
                     = a ^ exp / 10 ^ (18 * exp + 18) * 10 ^ -18
````

It seems to we just need to raise `a` to the power of `exp`, raise `10` to the power of `18 * exp + 18`, and do the division of the these two numbers, which requires one rounding only.

**Comment by devmannic** [comment]
Why not just implement traits from `num-traits` or reusing the macros that implement lots of these types of algorithms or reuse existing implementations altogether by converting to a different type do the operation then converting back?

  I don't think having a custom decimal implementation really makes sense apart from the existing lowest level functionality.

**Comment by r001** [comment]
> Why not just implement traits from `num-traits` or reusing the macros that implement lots of these types of algorithms or reuse existing implementations altogether by converting to a different type do the operation then converting back?
> 
> I don't think having a custom decimal implementation really makes sense apart from the existing lowest level functionality.

@devmannic There are two issues with floating point types:
1. The very big and very small problem: If you want to add to a very big floating point number a very small one, you have the problem that addition is not going to take place - SILENTLY - as the fractional part will not be able to represent the small addition. This leads to unexpected behaviour, and thus an exploit. Decimal addresses this, by only enabling as big numbers as the precision of the number can actually hold.
2. Binary representation problem: Floating point numbers represent numbers with a binary fractional part, and a binary exponent. Because of this, they can't even express the number of `1.7` correctly, because of having to convert them to binary, and thus the closest `1.69999...` will actually be stored. And this will also lead to: exploits.
See [this link](https://www.sciencedirect.com/topics/engineering/floating-point-number) for details.

**Comment by devmannic** [comment]
> > Why not just implement traits from `num-traits` or reusing the macros that implement lots of these types of algorithms or reuse existing implementations altogether by converting to a different type do the operation then converting back?
> > 
> > I don't think having a custom decimal implementation really makes sense apart from the existing lowest level functionality.
> 
> @devmannic There are two issues with floating point types:
> 1. The very big and very small problem: If you want to add to a very big floating point number a very small one, you have the problem that addition is not going to take place - SILENTLY - as the fractional part will not be able to represent the small addition. This leads to unexpected behaviour, and thus an exploit. Decimal addresses this, by only enabling as big numbers as the precision of the number can actually hold.
> 2. Binary representation problem: Floating point numbers represent numbers with a binary fractional part, and a binary exponent. Because of this, they can't even express the number of `1.7` correctly, because of having to convert them to binary, and thus the closest `1.69999...` will actually be stored. And this will also lead to: exploits.
> See [this link](https://www.sciencedirect.com/topics/engineering/floating-point-number) for details. 
> 

Yes I'm aware of these issues and agree we want solutions.  But my concern is in reimplementation of lots of operations and maintaining it let alone verifying it.  As an alternative but maybe slightly less performance approach see my implementation in defi/amm/bonding in the community repo for examples of how I solved these issues using num-rational and converting to/from Decimal.  It gets us a lot more for a lot less work in the short and long term.

https://github.com/radixdlt/community-scrypto-examples/blob/main/defi/amms/bonding/src/number.rs

**Comment by beemdvp** [comment]
Just wanted to say, thank you for looking into this as this was needed during foohack! And others have also been asking for this operation. Great job 👍

**Comment by r001** [comment]
This `powi()` works the following way: 
- It originates from the idea that `a ^ b = a * a * a ...` as b is integral, and a is also, we can do this. 
- Question: how to minimize multiplications? We use the fact that b = bit(0) * 2 ^ 0 + bit(1) * 2 ^ 1 + bit(2) * 2 ^ 2 + bit(3) * 2 ^ 3 + ... + bit(n) * 2 ^ n
 eg: if `bit(3)` of `b` is `1` then it means that we calculate `prod(n) = prod(n-1) * prod(n-1), prod(0) = a` and thus  `prod(3) = a ^ 8` and we multiply final result by this `res = res * prod(3)`. This way we only have to do around `log2(exponent)` rounds instead of `exponent` rounds.
- I have also renamed variables, and simplified code.
- Also fixed bug with negative power calculation.

I have also replaced a custom made BigInt -> Decimal function using trait `ToPrimitive` for `BigInt`.
@iamyulong @devmannic pls accept this one!

**Comment by r001** [comment]
I have rewritten the whole thing again to have a more concise syntax. The algo is based on:
https://en.wikipedia.org/wiki/Exponentiation_by_squaring

**Comment by iamyulong** [comment]
Awesome! We also need to use the checked `i32` arithmetic so overflow is captured for `release` build.

**Review by iamyulong — APPROVED** [review]
LGTM

**Pull request merged**